### PR TITLE
[5717] - Handle mapping for HESA for 2023/2024

### PR DESCRIPTION
**Pull request by d-a-v-e**
### Context

Each year, HESA changes the fields and codes in the HESA ITT collection. Often, this is to make it compatible with other data collections HESA does. Sometimes, this is because we have requested changes.

For the new academic year, we need to make sure we have mapped and handled all the changes to the HESA ITT collection 2023/24, so we import it correctly.

### Changes proposed in this pull request

See [the trello ticket](https://trello.com/c/7rRjkKl3/5717-handle-mapping-for-hesa-for-2023-2024) for a full rundown of additions and removals and feel free to run through with @d-a-v-e over a call.

### TODO

- [x] take copy of current collection
- [x] re-run import having cleared out any/all trainees with historical TRNs (either from a previous run during this testing or lsat years testing)
- [x] investigate/fix/understand the hiding of `hesa_trn_data` trainees from the trainee search/index page [slack thread](https://ukgovernmentdfe.slack.com/archives/C05M97JN6TE/p1692692838731059) • [trello ticket](https://trello.com/c/VXUwt2a1)
- [x] release minor version of `dfe-reference-data` [via this PR](https://github.com/DFE-Digital/dfe-reference-data/pull/82) and point to this new version in Register Gem file
- [x] Restore Staging workflow after merging #3512 

### TODO: Mapping Register => DQT

- [x] `QLAIM` - restore incorrectly removed `app/lib/hesa/code_sets/itt_qualification_aims.rb` [fixed here](https://github.com/DFE-Digital/register-trainee-teachers/pull/3465/commits/f122e50677583eec748cffa6b1353da029a00383)
- [x] `DEGCTRY` - confirm if anything needs changing on Register side. Currently we default to sending UK country code if the degree is UK based _[CONFIRMED IN SLACK HUDDLE - NOTHING TO BE CHANGED]_
- [x] `DEGCLSS` - known to be missing "distinction" could be missing more. Also relies on DTTP codeset mapping. ideally should be pulled out of DTTP namespace. [New mappings added here](https://github.com/DFE-Digital/register-trainee-teachers/pull/3465/commits/163cefbd3470f04ed49d5c2c13ff481ab2729185)
- [x] `SBJCA` - needs investigating further _...this is expected behaviour, [discussed here in slack](https://ukgovernmentdfe.slack.com/archives/C05M97JN6TE/p1692780174061639)_

### Important business

- [x] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [x] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?
- [x] Do we need to send any updates to DQT as part of the work in this PR?
- [x] Does this PR need an ADR?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

**Comment by d-a-v-e** [comment]
I've spotted an issue in `CourseSubjects` and need to fix. Will remove from draft once done.

**Comment by d-a-v-e** [comment]
[Handle mapping for HESA for 2023/2024](https://trello.com/c/7rRjkKl3/5717-handle-mapping-for-hesa-for-2023-2024)

File: app/services/trainees/create_from_hesa.rb [hunk 1]
```diff
@@ -232,7 +205,7 @@ def itt_end_date
     # HESA do not distinguish between the ITT start date and the trainee
     # start date, so we're setting both to the ITT start date.
     def trainee_start_date
-      itt_start_date
+      hesa_trainee[:trainee_start_date] || itt_start_date
```
Comment (defong) [inline app/services/trainees/create_from_hesa.rb]: I think the above comment can be removed as well, as it no longer just `itt_start_date`

File: app/services/trainees/create_from_hesa.rb [hunk 2]
```diff
@@ -222,7 +195,7 @@ def training_initiative
     # course, otherwise use HESA's commencement_date.  This is the start date
     # for trainees who have not transferred courses.
     def itt_start_date
-      hesa_trainee[:itt_commencement_date].presence || hesa_trainee[:commencement_date]
+      hesa_trainee[:itt_start_date]
```
Comment (defong) [inline app/services/trainees/create_from_hesa.rb:194]: I think the above comment can be removed as well, as it refers to `itt_commencement_date`

File: app/services/trainees/create_from_hesa.rb [hunk 3]
```diff
@@ -29,8 +29,9 @@ def call
         if trainee.save!
           create_degrees!
           create_placements!
+          store_application_choice_id!
```
Comment (defong) [inline app/services/trainees/create_from_hesa.rb:32]: Just need to follow up on 
1. How Hesa can have an `Apply application id`
2. Hesa 6 digits vs apply 10 characters, are we actually referring to the correct item?
image

File: app/services/trainees/create_from_hesa.rb [hunk 4]
```diff
@@ -148,7 +136,7 @@ def school_attributes
       end
 
       if hesa_trainee[:employing_school_urn].present?
-        if NOT_APPLICABLE_SCHOOL_URNS.include?(hesa_trainee[:lead_school_urn])
+        if NOT_APPLICABLE_SCHOOL_URNS.include?(hesa_trainee[:employing_school_urn])
```
Comment (defong) [inline app/services/trainees/create_from_hesa.rb:139]: Do we have a data discrepancy?

File: app/services/trainees/create_from_hesa.rb (hunk 3 above)
Comment (d-a-v-e) [reply app/services/trainees/create_from_hesa.rb:32]: @claire-hughez is this anything we need to consider on Register side?

File: app/services/trainees/create_from_hesa.rb (hunk 4 above)
Comment (d-a-v-e) [reply app/services/trainees/create_from_hesa.rb:139]: have raise with Claire and Sarah in the ticket

**Review by defong — APPROVED** [review]
As it stands all good, with just follow up questions to address, no showstoppers.

**Pull request merged**